File: ArchiveCapture/app/src/main/java/com/archiveprocessor/capture/ui/CaptureScreen.kt
package com.archiveprocessor.capture.ui

import android.Manifest
import android.content.pm.PackageManager
import android.graphics.BitmapFactory
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.camera.core.ImageCapture
import androidx.camera.core.ImageCaptureException
import androidx.camera.view.LifecycleCameraController
import androidx.camera.view.PreviewView
import androidx.compose.foundation.Image
import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.combinedClickable
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.slideOutVertically
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import androidx.core.content.ContextCompat
import androidx.lifecycle.compose.LocalLifecycleOwner
import com.archiveprocessor.capture.capture.CaptureViewModel
import com.archiveprocessor.capture.capture.CapturedItem
import com.archiveprocessor.capture.capture.GroupType
import com.archiveprocessor.capture.capture.UploadState
import java.io.File

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CaptureScreen(vm: CaptureViewModel) {
    val context = LocalContext.current
    val lifecycleOwner = LocalLifecycleOwner.current
    var hasCam by remember {
        mutableStateOf(
            ContextCompat.checkSelfPermission(context, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED
        )
    }
    val permLauncher = rememberLauncherForActivityResult(ActivityResultContracts.RequestPermission()) { hasCam = it }
    LaunchedEffect(Unit) { if (!hasCam) permLauncher.launch(Manifest.permission.CAMERA) }

    val controller = remember { LifecycleCameraController(context) }
    LaunchedEffect(hasCam) { if (hasCam) controller.bindToLifecycle(lifecycleOwner) }

    // The strip shows only current in-flight work — document pages (queued/transferring) plus any failed
    // markers to retry. Confirmed uploads leave the phone, so images never pile up here.
    val strip = vm.items.filter { it.type == GroupType.DOCUMENT || it.state == UploadState.PENDING || it.state == UploadState.FAILED }
    // Auto-scroll the strip so the newest capture stays in view.
    val listState = rememberLazyListState()
    LaunchedEffect(strip.size) {
        if (strip.isNotEmpty()) listState.animateScrollToItem(strip.size - 1)
    }

    var showClearConfirm by remember { mutableStateOf(false) }
    var showRepairConfirm by remember { mutableStateOf(false) }

    Column(Modifier.fillMaxSize().background(Color.Black)) {
        // Camera preview — top region only, letterboxed (FIT_CENTER) on black.
        Box(Modifier.fillMaxWidth().weight(1f), contentAlignment = Alignment.Center) {
            if (hasCam) {
                AndroidView(
                    factory = { ctx ->
                        PreviewView(ctx).apply {
                            this.controller = controller
                            scaleType = PreviewView.ScaleType.FIT_CENTER
                        }
                    },
                    modifier = Modifier.fillMaxSize()
                )
            } else {
                Text("Camera permission needed to capture.", color = Color.White)
            }
        }

        // Controls region (below the preview).
        Column(
            Modifier.fillMaxWidth().background(Color(0xFF141414)).padding(horizontal = 12.dp, vertical = 10.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            // Connection status + Re-pair. Once paired the app goes straight to this screen, so this is
            // the only way back to the QR scanner — e.g. to switch a USB-paired phone (host 127.0.0.1)
            // over to Wi-Fi. Captured photos are kept and upload after reconnecting.
            Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                Text(vm.endpoint?.let { "Connected · ${it.name}" } ?: "Not connected",
                     color = Color(0xFF8E8E93), style = MaterialTheme.typography.labelMedium)
                Spacer(Modifier.weight(1f))
                TextButton(onClick = { showRepairConfirm = true }) { Text("Re-pair", color = Color.White) }
            }

            // End segment — above the photos and away from the shutter, to avoid accidental taps.
            Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                if (vm.items.isNotEmpty()) {
                    TextButton(onClick = { showClearConfirm = true }) { Text("Clear", color = Color.White) }
                }
                if (vm.items.any { it.state == UploadState.FAILED }) {
                    TextButton(onClick = { vm.retryFailed() }) { Text("Retry", color = Color.White) }
                }
                Spacer(Modifier.weight(1f))
                Button(onClick = { vm.finishDocumentSegment() }) { Text("End segment") }
                Spacer(Modifier.weight(1f))
                TextButton(onClick = { vm.finishSession() }) { Text("Finish", color = Color.White) }
            }

            // Transfer feedback: segments/markers fly to the Mac; images don't accumulate on the phone.
            val uploading = vm.items.count { it.state == UploadState.UPLOADING }
            if (vm.transferFlash != null || uploading > 0) {
                Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    AnimatedVisibility(visible = vm.transferFlash != null, enter = fadeIn(), exit = fadeOut()) {
                        Text("⤴ ${vm.transferFlash ?: ""}", color = Color(0xFF34C759), style = MaterialTheme.typography.labelLarge)
                    }
                    Spacer(Modifier.weight(1f))
                    if (uploading > 0) Text("Transferring $uploading…", color = Color(0xFFFFCC00), style = MaterialTheme.typography.labelMedium)
                }
            }

            // Status line — surfaces capture errors (a failed shutter can't be silent; archival photos
            // can't be re-taken) and the recovered-segment prompt after a restart.
            if (vm.statusMessage.isNotEmpty()) {
                Text(vm.statusMessage, color = Color.White, style = MaterialTheme.typography.bodySmall,
                     modifier = Modifier.fillMaxWidth())
            }

            // Current segment (auto-scrolling; tap a page to toggle its P10 override). Confirmed pages
            // animate out as they reach the Mac, so the strip reflects only what's still transferring.
            if (strip.isNotEmpty()) {
                LazyRow(state = listState, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    items(strip, key = { it.id }) { item ->
                        AnimatedVisibility(
                            visible = item.state != UploadState.UPLOADED,
                            exit = fadeOut() + slideOutVertically { -it }
                        ) {
                            Thumb(
                                item = item,
                                isSelected = vm.selectedItemId == item.id && !vm.armed,
                                isArmed = vm.selectedItemId == item.id && vm.armed,
                                onTap = { vm.tapItem(item.id) },
                                onLongPress = { vm.toggleP10(item.id) }
                            )
                        }
                    }
                }
            }

            // Capture row: Box (red) · white shutter · Folder (purple). Shutter is at the very bottom.
            Row(
                Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.SpaceEvenly
            ) {
                Button(
                    onClick = {
                        if (vm.selectedItemId != null) vm.reclassifySelected(GroupType.BOX)
                        else takePicture(context, controller, vm) { vm.captureMarker(it, GroupType.BOX) }
                    },
                    colors = ButtonDefaults.buttonColors(containerColor = Color(0xFFD32F2F), contentColor = Color.White)
                ) { Text("Box") }

                Button(
                    onClick = { takePicture(context, controller, vm) { vm.addDocumentPhoto(it) } },
                    shape = CircleShape,
                    colors = ButtonDefaults.buttonColors(containerColor = Color.White),
                    modifier = Modifier.size(76.dp).border(3.dp, Color.LightGray, CircleShape)
                ) { }

                Button(
                    onClick = {
                        if (vm.selectedItemId != null) vm.reclassifySelected(GroupType.FOLDER)
                        else takePicture(context, controller, vm) { vm.captureMarker(it, GroupType.FOLDER) }
                    },
                    colors = ButtonDefaults.buttonColors(containerColor = Color(0xFF7B1FA2), contentColor = Color.White)
                ) { Text("Folder") }
            }
        }
    }

    // Segment tag sheet.
    if (vm.pendingTagGroupId != null) {
        val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
        ModalBottomSheet(onDismissRequest = { vm.cancelTagSheet() }, sheetState = sheetState) {
            SegmentTagSheet(recentYears = vm.recentYears()) { p, y, m -> vm.applyTagsAndContinue(p, y, m) }
        }
    }

    // Confirm before clearing (deletes photos from the phone — destructive).
    if (showClearConfirm) {
        AlertDialog(
            onDismissRequest = { showClearConfirm = false },
            title = { Text("Clear all photos?") },
            text = { Text("This permanently deletes all ${vm.items.size} captured photo(s) from this phone and cannot be undone.") },
            confirmButton = {
                TextButton(onClick = { vm.clearSession(); showClearConfirm = false }) {
                    Text("Clear", color = Color(0xFFD32F2F))
                }
            },
            dismissButton = { TextButton(onClick = { showClearConfirm = false }) { Text("Cancel") } }
        )
    }

    // Re-pair: disconnect and return to the pairing screen (e.g. to move from USB to Wi-Fi). Non-destructive.
    if (showRepairConfirm) {
        AlertDialog(
            onDismissRequest = { showRepairConfirm = false },
            title = { Text("Re-pair with a Mac?") },
            text = { Text("Disconnects from ${vm.endpoint?.name ?: "the Mac"} and returns to the pairing screen so you can scan a QR — e.g. to switch from USB to Wi-Fi. Any captured photos are kept and upload once you reconnect.") },
            confirmButton = { TextButton(onClick = { vm.disconnect(); showRepairConfirm = false }) { Text("Re-pair") } },
            dismissButton = { TextButton(onClick = { showRepairConfirm = false }) { Text("Cancel") } }
        )
    }
}

private fun takePicture(
    context: android.content.Context,
    controller: LifecycleCameraController,
    vm: CaptureViewModel,
    onSaved: (File) -> Unit
) {
    val file = vm.newCaptureFile()
    val opts = ImageCapture.OutputFileOptions.Builder(file).build()
    controller.takePicture(
        opts,
        ContextCompat.getMainExecutor(context),
        object : ImageCapture.OnImageSavedCallback {
            override fun onImageSaved(results: ImageCapture.OutputFileResults) { onSaved(file) }
            override fun onError(exception: ImageCaptureException) {
                // A failed capture must NOT be silent — an archival page can't be re-taken. Remove any
                // partial file and surface the failure so the operator re-shoots.
                runCatching { file.delete() }
                vm.reportCaptureError("Capture failed — please retake. (${exception.message ?: "camera error"})")
            }
        }
    )
}

@OptIn(ExperimentalFoundationApi::class)
@Composable
private fun Thumb(
    item: CapturedItem,
    isSelected: Boolean,
    isArmed: Boolean,
    onTap: () -> Unit,
    onLongPress: () -> Unit
) {
    val bmp: ImageBitmap? = remember(item.file.path) { decodeThumb(item.file.path) }
    val stateColor = when (item.state) {
        UploadState.UPLOADED -> Color(0xFF34C759)
        UploadState.UPLOADING -> Color(0xFFFFCC00)
        UploadState.FAILED -> Color(0xFFFF3B30)
        UploadState.PENDING -> Color.Gray
    }
    val isP10 = item.priority == "P10"
    val ring = when {
        isArmed -> Color(0xFFFF3B30)     // red: tap again to delete
        isSelected -> Color(0xFF0A84FF)  // blue: selected
        isP10 -> Color(0xFFFFD60A)       // gold: P10
        else -> null
    }
    Box(
        Modifier.size(64.dp).clip(RoundedCornerShape(6.dp)).background(Color.DarkGray)
            .then(if (ring != null) Modifier.border(3.dp, ring, RoundedCornerShape(6.dp)) else Modifier)
            .combinedClickable(onClick = onTap, onLongClick = onLongPress)
    ) {
        if (bmp != null) {
            Image(bitmap = bmp, contentDescription = null, contentScale = ContentScale.Crop, modifier = Modifier.fillMaxSize())
        }
        Box(Modifier.align(Alignment.BottomStart).padding(3.dp).size(10.dp).clip(CircleShape).background(stateColor))
        if (isP10 && !isArmed) {
            Box(Modifier.align(Alignment.TopEnd).padding(2.dp).clip(RoundedCornerShape(3.dp)).background(Color.Red)) {
                Text("P10", color = Color.White, style = MaterialTheme.typography.labelSmall, modifier = Modifier.padding(horizontal = 2.dp))
            }
        }
        if (isArmed) {
            Box(Modifier.fillMaxSize().background(Color.Black.copy(alpha = 0.45f)), contentAlignment = Alignment.Center) {
                Text("✕", color = Color.White, style = MaterialTheme.typography.headlineMedium)
            }
        }
    }
}

/** Downsampled thumbnail decode so the strip never loads full-resolution camera JPEGs. */
private fun decodeThumb(path: String, target: Int = 200): ImageBitmap? = try {
    val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
    BitmapFactory.decodeFile(path, bounds)
    var sample = 1
    while (bounds.outWidth / sample > target || bounds.outHeight / sample > target) sample *= 2
    val opts = BitmapFactory.Options().apply { inSampleSize = sample }
    BitmapFactory.decodeFile(path, opts)?.asImageBitmap()
} catch (e: Exception) {
    null
}

File: ArchiveCaptureiOS/Sources/ArchiveCaptureiOS/UI/CaptureScreen.swift
import SwiftUI
import ImageIO
import UIKit

/// Main capture UI: camera preview + Box/shutter/Folder, End-segment, and a strip that shows only the
/// current in-flight work — confirmed uploads leave the phone (photos transfer to the Mac in segments).
struct CaptureScreen: View {
    @ObservedObject var vm: CaptureViewModel
    @StateObject private var camera = CameraController()
    @State private var showClearConfirm = false
    @State private var showRepairConfirm = false
    @State private var isCapturing = false

    /// Current in-flight work: document pages, plus any PENDING/FAILED marker needing attention.
    private var strip: [CapturedItem] {
        vm.items.filter { $0.type == .document || $0.state == .pending || $0.state == .failed }
    }

    var body: some View {
        VStack(spacing: 0) {
            ZStack {
                Color.black
                if camera.authorized {
                    CameraPreview(session: camera.session)
                } else {
                    VStack(spacing: 12) {
                        Text("Camera permission needed to capture.").foregroundStyle(.white)
                        if camera.accessDenied {
                            Button("Open Settings") {
                                if let url = URL(string: UIApplication.openSettingsURLString) {
                                    UIApplication.shared.open(url)
                                }
                            }
                            .buttonStyle(.borderedProminent)
                        }
                    }
                }
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity)

            VStack(spacing: 8) {
                // Connection + Re-pair: once paired the app opens straight to this screen, so this is the
                // way back to the QR scanner (switch Macs/networks). Captured photos are kept meanwhile.
                HStack {
                    Text(vm.endpoint.map { "Connected · \($0.name)" } ?? "Not connected")
                        .font(.caption).foregroundStyle(.secondary)
                    Spacer()
                    Button("Re-pair") { showRepairConfirm = true }.foregroundStyle(.white)
                }
                HStack {
                    if !vm.items.isEmpty { Button("Clear") { showClearConfirm = true }.foregroundStyle(.white) }
                    if vm.items.contains(where: { $0.state == .failed }) { Button("Retry") { vm.retryFailed() }.foregroundStyle(.white) }
                    Spacer()
                    Button("End segment") { vm.finishDocumentSegment() }.buttonStyle(.borderedProminent)
                    Spacer()
                    Button("Finish") { vm.finishSession() }.foregroundStyle(.white)
                }

                let uploading = vm.items.filter { $0.state == .uploading }.count
                if vm.transferFlash != nil || uploading > 0 {
                    HStack(spacing: 8) {
                        if let flash = vm.transferFlash { Text("⤴ \(flash)").foregroundStyle(.green).font(.callout) }
                        Spacer()
                        if uploading > 0 { Text("Transferring \(uploading)…").foregroundStyle(.yellow).font(.caption) }
                    }
                }

                if !strip.isEmpty {
                    ScrollView(.horizontal, showsIndicators: false) {
                        HStack(spacing: 8) {
                            ForEach(strip) { item in
                                Thumb(item: item,
                                      selected: vm.selectedItemId == item.id && !vm.armed,
                                      armed: vm.selectedItemId == item.id && vm.armed)
                                    .onTapGesture { vm.tapItem(item.id) }
                                    .onLongPressGesture { vm.toggleP10(item.id) }
                                    .transition(.opacity.combined(with: .move(edge: .top)))
                            }
                        }
                        .padding(.vertical, 2)
                    }
                    .frame(height: 70)
                    .animation(.easeInOut(duration: 0.25), value: strip.map(\.id))
                }

                HStack {
                    Button { capture(.box) } label: { Text("Box").frame(width: 66, height: 44) }
                        .background(Color(red: 0.83, green: 0.18, blue: 0.18)).foregroundStyle(.white)
                        .clipShape(RoundedRectangle(cornerRadius: 8))
                    Spacer()
                    Button { capture(.document) } label: {
                        Circle().fill(.white).frame(width: 68, height: 68).overlay(Circle().stroke(.gray, lineWidth: 3))
                    }
                    Spacer()
                    Button { capture(.folder) } label: { Text("Folder").frame(width: 66, height: 44) }
                        .background(Color(red: 0.48, green: 0.12, blue: 0.64)).foregroundStyle(.white)
                        .clipShape(RoundedRectangle(cornerRadius: 8))
                }
                .disabled(isCapturing)

                if !vm.statusMessage.isEmpty {
                    Text(vm.statusMessage).font(.caption2).foregroundStyle(.white.opacity(0.6))
                }
            }
            .padding(12)
            .background(Color(white: 0.08))
        }
        .background(Color.black.ignoresSafeArea())
        .onAppear { camera.start() }
        .onDisappear { camera.stop() }
        .sheet(isPresented: Binding(get: { vm.pendingTagGroupId != nil },
                                    set: { if !$0 { vm.cancelTagSheet() } })) {
            SegmentTagSheet(recentYears: vm.recentYears) { p, y, m in
                vm.applyTagsAndContinue(priority: p, year: y, month: m)
            }
        }
        .confirmationDialog("Clear all photos?", isPresented: $showClearConfirm, titleVisibility: .visible) {
            Button("Clear (\(vm.items.count))", role: .destructive) { vm.clearSession() }
            Button("Cancel", role: .cancel) {}
        } message: {
            Text("This permanently deletes captured photos still on this phone and can't be undone.")
        }
        .confirmationDialog("Re-pair with a Mac?", isPresented: $showRepairConfirm, titleVisibility: .visible) {
            Button("Re-pair") { vm.disconnect() }
            Button("Cancel", role: .cancel) {}
        } message: {
            Text("Disconnects from \(vm.endpoint?.name ?? "the Mac") and returns to the pairing screen so you can scan a QR (e.g. to switch Macs or networks). Captured photos are kept and upload once you reconnect.")
        }
        .alert("Photo not saved", isPresented: Binding(
            get: { vm.captureError != nil },
            set: { if !$0 { vm.captureError = nil } }
        )) {
            Button("OK", role: .cancel) { vm.captureError = nil }
        } message: {
            Text(vm.captureError ?? "")
        }
    }

    /// Reclassify a selected page into a Box/Folder if one is selected; otherwise take a new photo.
    private func capture(_ type: GroupType) {
        if type != .document, vm.selectedItemId != nil { vm.reclassifySelected(type); return }
        guard !isCapturing else { return }
        isCapturing = true
        camera.capturePhoto { data in
            isCapturing = false
            guard let data else { return }
            guard let url = vm.persistCapturedJPEG(data) else { return }   // surfaces a blocking alert on failure
            if type == .document { vm.addDocumentPhoto(url) } else { vm.captureMarker(url, type: type) }
        }
    }
}

/// A capture strip thumbnail with an upload-state dot; loads a downsampled image off the main thread.
private struct Thumb: View {
    let item: CapturedItem
    let selected: Bool
    let armed: Bool
    @State private var image: UIImage?

    var body: some View {
        ZStack(alignment: .bottomLeading) {
            if let image { Image(uiImage: image).resizable().scaledToFill() }
            else { Color.gray.opacity(0.4) }
            Circle().fill(stateColor).frame(width: 10, height: 10).padding(3)
            if armed {
                Color.black.opacity(0.45)
                Image(systemName: "xmark").foregroundStyle(.white).font(.headline)
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
            }
        }
        .frame(width: 64, height: 64)
        .clipped()
        .clipShape(RoundedRectangle(cornerRadius: 6))
        .overlay(RoundedRectangle(cornerRadius: 6).stroke(ring, lineWidth: ring == .clear ? 0 : 3))
        .task(id: item.fileURL) { image = await Self.loadThumb(item.fileURL) }
    }

    private var stateColor: Color {
        switch item.state {
        case .uploaded: return .green
        case .uploading: return .yellow
        case .failed: return .red
        case .pending: return .gray
        }
    }
    private var ring: Color {
        if armed { return .red }
        if selected { return .blue }
        if item.priority == "P10" { return .yellow }
        return .clear
    }

    private static func loadThumb(_ url: URL, maxPixel: Int = 180) async -> UIImage? {
        await Task.detached(priority: .userInitiated) {
            guard let src = CGImageSourceCreateWithURL(url as CFURL, nil),
                  let cg = CGImageSourceCreateThumbnailAtIndex(src, 0, [
                      kCGImageSourceCreateThumbnailFromImageAlways: true,
                      kCGImageSourceThumbnailMaxPixelSize: maxPixel,
                      kCGImageSourceCreateThumbnailWithTransform: true
                  ] as CFDictionary) else { return nil }
            return UIImage(cgImage: cg)
        }.value
    }
}
